client.go

```diff
@@ -210,7 +203,24 @@ func (c *Client) connect() (err error) {
 	authComplete := make(chan error)
 
 	go c.handleRawServerMessages(authComplete)
-	go c.handleOpcodes(authComplete)
+	go func() {
+		c.handleOpcodes(authComplete)
+
+		// we write to IncomingResponses only from one place:
+		// * c.handleOpcodes
+		// and we also read from it in:
+		// * c.client.SendRequest
+		// thus the `close` must happen only after c.handleOpcodes finished,
+		// and is desired to happen after c.client.SendRequest will stop
+		// using the channel as well (to avoid handling nil Responses).
+		//
+		// This line right here:
+		// * it is after c.handleOpcodes.
+		// * this place is reachable only after c.client.Opcodes is closed,
+		//   which is possible only when c.client.Disconnected is closed,
+		//   which means c.client.SendRequest would not write into c.client.IncomingResponses.
```

Comment:
Sorry I'm having a hard time following the reasoning here. Previously `c.markDisconnected()` closed the opcodes channel first, at which point the loop in `c.handleOpcodes()` terminates, and since that's the only place where `c.client.IncomingResponses` is written to, that channel is closed thereafter:

https://github.com/andreykaipov/goobs/blob/bd2c39db013ff66b2c8703263c11268fa475d8aa/client.go#L141-L145

Can you help me understand what's different between this approach versus the above disconnection logic? I know things with concurrency are hard to reproduce, but a code snippet or something would help so I can try it myself to better understand the behaviour.

Reply:
> that channel is closed thereafter

If you mean that `close(c.client.IncomingResponses) ` is invoked after `close(c.client.Opcodes) ` then it is not good enough, because those are two separate goroutines (the one that closes the channels, and the other that consumes/writes (/to) the channels), so it is essentially a race condition.

> Can you help me understand what's different between this approach versus the above disconnection logic?

The old logic is:
* Start closing channels in one goroutine, while working with these channels in other goroutines. as a result we can do a lot of bad stuff, like writing to `IncomingResponses` after it is closed (because the writing routine checked `OpCodes` and it was not closed and proceeded to work with `IncomingResponses`, but in the meanwhile it was closed); and there are other examples.

The new logic is:
* Close `c.client.Disconnected` and `c.client.OpCodes` while locking `c.client.mutex` to avoid race condition in `SendRequest` (because `SendRequest` checks the `Disconnected` and proceeds with sending via `OpCodes` while holding `mutex`). The race condition here is now fixed.
* Close `c.client.IncomingResponses`  only after the writer (`handleOpcodes`) to this channel have finished. Thus race condition here is now also fixed.
* Close `c.client.IncomingResponses` only after `SendRequest` stopped reading from `c.client.IncomingResponses`, otherwise it will read a `nil`-valued `*opcodes.RequestResponse` and panic in line `if response.ID != id {`. And this order is achieved because `handleOpcodes` is finished only when `c.client.Opcodes` is closed, which happens only when `c.client.Disconnect` is closed, and thus `SendRequest` will not get into a branch where it reads from `c.client.IncomingResponses`, and no race condition here is guaranteed by the mutex (one more reason why we moved closure of some channels into function `Close` where we can lock `mutex`, which is a private field).

> I know things with concurrency are hard to reproduce, but a code snippet or something would help so I can try it myself to better understand the behaviour.

Here is an example: https://play.golang.com/p/4yy8Bm7IVDS
Please let me know if it is not descriptive enough :)

Reply:
Sorry I didn't have time to check this out on the weekend. I think understand this better than I did before, but my knowledge is definitely lacking in there area so I really appreciate you walking me through it! Don't want to hold this up any longer. I hope this fixes https://github.com/andreykaipov/goobs/issues/182 :)